File: packages/skill-eval/src/providers/openai.ts
import OpenAI from "openai";
import { BaseProvider, buildChatResponse, normalizeChatParams } from "./base.js";
import { zodToJsonSchema } from "./schema.js";
import type { ChatParams, ChatResponse, ToolCall, ToolDefinition } from "./types.js";

function formatTools(tools: ToolDefinition[]): OpenAI.Chat.Completions.ChatCompletionTool[] {
	return tools.map((t) => ({
		type: "function" as const,
		function: {
			name: t.name,
			description: t.description,
			parameters: zodToJsonSchema(t.parameters),
		},
	}));
}

const STOP_MAP: Record<string, ChatResponse["stopReason"]> = {
	stop: "end",
	tool_calls: "tool_use",
	length: "max_tokens",
};

export class OpenAIProvider extends BaseProvider {
	readonly name = "openai";
	private readonly client: OpenAI;

	constructor(apiKey: string, model: string) {
		super(model);
		this.client = new OpenAI({ apiKey });
	}

	async chat(params: ChatParams): Promise<ChatResponse> {
		const start = Date.now();
		const { maxTokens, temperature } = normalizeChatParams(params);

		const messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [
			{ role: "system", content: params.system },
		];

		for (const m of params.messages) {
			if (m.role === "tool_result") {
				messages.push({
					role: "tool",
					tool_call_id: m.toolCallId ?? "",
					content: m.content,
				});
			} else if (m.role === "assistant" && m.toolCalls?.length) {
				messages.push({
					role: "assistant",
					content: m.content || null,
					tool_calls: m.toolCalls.map((tc) => ({
						id: tc.id,
						type: "function" as const,
						function: { name: tc.name, arguments: JSON.stringify(tc.arguments) },
					})),
				});
			} else {
				messages.push({
					role: m.role as "user" | "assistant",
					content: m.content,
				});
			}
		}

		const response = await this.client.chat.completions.create({
			model: this.modelId,
			messages,
			tools: params.tools?.length ? formatTools(params.tools) : undefined,
			temperature,
			max_completion_tokens: maxTokens,
		});

		const choice = response.choices[0];
		const toolCalls: ToolCall[] = (choice.message.tool_calls ?? [])
			.filter((tc): tc is Extract<typeof tc, { type: "function" }> => tc.type === "function")
			.map((tc) => ({
				id: tc.id,
				name: tc.function.name,
				arguments: JSON.parse(tc.function.arguments),
			}));

		return buildChatResponse({
			content: choice.message.content ?? "",
			toolCalls,
			inputTokens: response.usage?.prompt_tokens ?? 0,
			outputTokens: response.usage?.completion_tokens ?? 0,
			stopReason: STOP_MAP[choice.finish_reason ?? "stop"] ?? "end",
			latencyMs: Date.now() - start,
		});
	}
}

File: packages/skill-eval-python/src/skill_eval/providers/openai.py
from __future__ import annotations

import asyncio
import json
import time
from typing import Any

from openai import OpenAI

from skill_eval.providers.base import BaseProvider, build_chat_response, normalize_chat_params
from skill_eval.providers.types import ChatParams, ChatResponse, ToolCall, ToolDefinition

STOP_MAP: dict[str, str] = {
    "stop": "end",
    "tool_calls": "tool_use",
    "length": "max_tokens",
}


def _get_tool_schema(tool: ToolDefinition) -> dict[str, Any]:
    params = tool.parameters
    if isinstance(params, type) and hasattr(params, "model_json_schema"):
        return params.model_json_schema()
    if isinstance(params, dict) and params:
        return params
    return {"type": "object", "properties": {}}


def _format_tools(tools: list[ToolDefinition]) -> list[dict[str, Any]]:
    return [
        {
            "type": "function",
            "function": {
                "name": t.name,
                "description": t.description,
                "parameters": _get_tool_schema(t),
            },
        }
        for t in tools
    ]


class OpenAIProvider(BaseProvider):
    def __init__(self, api_key: str, model: str) -> None:
        super().__init__(model)
        self._client = OpenAI(api_key=api_key)

    @property
    def name(self) -> str:
        return "openai"

    async def chat(self, params: ChatParams) -> ChatResponse:
        start = time.perf_counter()
        max_tokens, temperature = normalize_chat_params(params)

        messages: list[dict[str, Any]] = [
            {"role": "system", "content": params.system}
        ]

        for m in params.messages:
            if m.role == "tool_result":
                messages.append(
                    {
                        "role": "tool",
                        "tool_call_id": m.tool_call_id or "",
                        "content": m.content,
                    }
                )
            elif m.role == "assistant" and m.tool_calls:
                messages.append(
                    {
                        "role": "assistant",
                        "content": m.content or None,
                        "tool_calls": [
                            {
                                "id": tc.id,
                                "type": "function",
                                "function": {
                                    "name": tc.name,
                                    "arguments": json.dumps(tc.arguments),
                                },
                            }
                            for tc in m.tool_calls
                        ],
                    }
                )
            else:
                messages.append({"role": m.role, "content": m.content})

        def _create() -> Any:
            return self._client.chat.completions.create(
                model=self.model_id,
                messages=messages,
                tools=_format_tools(params.tools) if params.tools else None,
                temperature=temperature,
                max_completion_tokens=max_tokens,
            )

        response = await asyncio.to_thread(_create)

        choice = response.choices[0]
        tool_calls: list[ToolCall] = []
        for tc in choice.message.tool_calls or []:
            if tc.type == "function":
                tool_calls.append(
                    ToolCall(
                        id=tc.id,
                        name=tc.function.name,
                        arguments=json.loads(tc.function.arguments or "{}"),
                    )
                )

        latency_ms = (time.perf_counter() - start) * 1000
        stop_reason = STOP_MAP.get(choice.finish_reason or "stop", "end")

        return build_chat_response(
            content=choice.message.content or "",
            tool_calls=tool_calls,
            input_tokens=response.usage.prompt_tokens if response.usage else 0,
            output_tokens=response.usage.completion_tokens if response.usage else 0,
            stop_reason=stop_reason,
            latency_ms=latency_ms,
        )
